mqtt/mqtt-bridge/tests/integration.rs

```diff
@@ -122,6 +123,89 @@ async fn send_message_upstream_downstream() {
     local_client.shutdown().await;
 }
 
+#[ignore = "This test can be activated once RingBuffer is merged in."]
+#[tokio::test]
+async fn send_message_upstream_with_crash_is_lossless() {
+    let subs = vec![Direction::Out(TopicRule::new(
+        "temp/#".into(),
+        Some("to".into()),
+        Some("upstream".into()),
+    ))];
+
+    let (mut local_server_handle, _, mut upstream_server_handle, _) =
+        common::setup_brokers(AllowAll, AllowAll);
+
+    let mut local_client = TestClientBuilder::new(local_server_handle.address())
+        .with_client_id(ClientId::IdWithExistingSession("local_client".into()))
+        .build();
+
+    let mut upstream_client = TestClientBuilder::new(upstream_server_handle.address())
+        .with_client_id(ClientId::IdWithExistingSession("upstream_client".into()))
+        .build();
+    {
+        let (controller_handle, controller_task) = common::setup_bridge_controller(
+            local_server_handle.address(),
+            upstream_server_handle.tls_address().unwrap(),
+            subs.clone(),
+            "send_message_upstream_with_crash_is_lossless",
+        )
+        .await;
+
+        local_client
+            .subscribe("downstream/filter/#", QoS::AtLeastOnce)
+            .await;
+
+        // wait to receive subscription ack
+        local_client.subscriptions().next().await;
+
+        upstream_client
+            .subscribe("upstream/temp/#", QoS::AtLeastOnce)
+            .await;
+
+        // wait to receive subscription ack
+        upstream_client.subscriptions().next().await;
+
+        // send upstream
+        local_client
+            .publish_qos1("to/temp/1", "from local", false)
+            .await;
+
+        assert_matches!(
+            upstream_client.publications().next().await,
+            Some(ReceivedPublication{payload, .. }) if payload == Bytes::from("from local")
+        );
+
+        // send upstream but disconnect afterwards
+        local_client
+            .publish_qos1("to/temp/1", "from local again", false)
+            .await;
+
+        controller_handle.shutdown();
```

Comment:
So, there is a catch here... the main problem is that clients (`TestClient`) that you are using are "active". It means they do work in the background task and fetch/push messages as fast as they can. They also hide some protocol complexity by taking care of acks, etc.

So, in this case there is no guarantee that the upstream client did not receive the message prior to shutdown. It might have received that and kept it in it's local memory. Later on line 195 you could just get it from local memory and not the bridge as you might think.

What you need to do is to use `PacketStream` instead. `PacketStream` is "passive", and performs work only when asked. It also allows to have more control over the packets, like you're responsible for sending acks yourself with `PacketStream`. So, you can make one of the client to be `PacketStream` and don't send ack when you receive a publication. That will keep it in the bridge. 

Then you can restart the bridge, reconnect the client and see if message is being re-delivered from the storage

Reply:
Yeah, I found this out the hard way today :P I saw another test that just shutdowns the controller handle and the associated task and then restarts. I have now copied that version